packages/@aws-cdk/aws-eks/lib/cluster-resource.ts

```diff
@@ -32,51 +30,66 @@ export class ClusterResource extends Construct {
    * that gets administrator privilages on the cluster (`system:masters`), and
    * will be able to issue `kubectl` commands against it.
    */
-  public readonly creationRole: iam.IRole;
+  private readonly creationRole: iam.Role;
+  private readonly trustedPrincipals: string[] = [];
 
   constructor(scope: Construct, id: string, props: CfnClusterProps) {
     super(scope, id);
 
-    // each cluster resource will have it's own lambda handler since permissions
-    // are scoped to this cluster and related resources like it's role
-    const handler = new lambda.Function(this, 'ResourceHandler', {
-      code: lambda.Code.fromAsset(path.join(__dirname, 'cluster-resource')),
-      runtime: lambda.Runtime.PYTHON_3_7,
-      handler: 'index.handler',
-      timeout: Duration.minutes(15),
-      memorySize: 512,
-      layers: [ KubectlLayer.getOrCreate(this) ],
-    });
+    const provider = ClusterResourceProvider.getOrCreate(this);
 
     if (!props.roleArn) {
       throw new Error(`"roleArn" is required`);
     }
 
-    // since we don't know the cluster name at this point, we must give this role star resource permissions
-    handler.addToRolePolicy(new iam.PolicyStatement({
-      actions: [ 'eks:CreateCluster', 'eks:DescribeCluster', 'eks:DeleteCluster', 'eks:UpdateClusterVersion' ],
-      resources: [ '*' ]
-    }));
+    // the role used to create the cluster. this becomes the administrator role
+    // of the cluster.
+    this.creationRole = new iam.Role(this, 'CreationRole', {
+      assumedBy: new iam.CompositePrincipal(...provider.roles.map(x => new iam.ArnPrincipal(x.roleArn)))
+    });
 
     // the CreateCluster API will allow the cluster to assume this role, so we
     // need to allow the lambda execution role to pass it.
-    handler.addToRolePolicy(new iam.PolicyStatement({
+    this.creationRole.addToPolicy(new iam.PolicyStatement({
       actions: [ 'iam:PassRole' ],
       resources: [ props.roleArn ]
     }));
 
+    // since we don't know the cluster name at this point, we must give this role star resource permissions
```

Comment:
We know the cluster name if the user specified it...

Reply:
That's a good point, I can use it if it is explicitly specified.

Reply:
Dobne